internal/service/cluster/cluster.go

```diff
@@ -387,7 +387,62 @@ func (s *Service) Run(ctx context.Context, host service.Host) error {
 	return nil
 }
 
-func (s *Service) getPeers() ([]string, error) {
+func (s *Service) notifyComponentsOfClusterChanges(ctx context.Context, limiter *rate.Limiter, host service.Host) {
+	tracer := s.tracer.Tracer("")
+	spanCtx, span := tracer.Start(ctx, "NotifyClusterChange", trace.WithSpanKind(trace.SpanKindInternal))
+
+	// Limit how often we notify components about peer changes. At start up we may receive N updates in a period
+	// of less than one second. This leads to a lot of unnecessary processing.
+	_, spanWait := tracer.Start(spanCtx, "RateLimitWait", trace.WithSpanKind(trace.SpanKindInternal))
+	if err := limiter.Wait(ctx); err != nil {
```

Comment:
Is there a risk here that we block notification of not ready -> ready due to the limiter?

Reply:
So in this case of limiter, we will block instead of aborting the operation when there's no budget left in the limiter. Blocking for up to 1s means that the `Ready()` when called directly from components will be stale (so, for example, cluster may not be Ready(), even though a call to Peers returns enough peers). After limiter unblocks we update the `Ready()` and notify all the components right away, at which point they should have consistent view. 

I don't see a problem with this delay really. The godocs say "Always use Ready to verify before assigning work to instance." and our components currently follow this correctly.

Reply:
Okay, I don't think it's a problem, but it also doesn't really cost a thing if we move that state update before the limiter. I've done this now with a comment.

```
	// Update Ready() state of cluster service that components use. Doing it before the limiter to reduce the time
	// during which the components' view of cluster is not fully consistent (e.g. cluster not Ready() even though 
	// the number of peers is sufficient). Calls to `updateReadyState()` will still be effectively rate-limited
	// because only one goroutine performs the notification.
	s.alloyCluster.updateReadyState()

```

Reply:
After 1 second, which is this limiter's configured wait time. Note that if there's no budget left in the limiter, the `Wait` will block until budget is available and then it will continue. It's different then calling `Allow()` and aborting, for example. 

See limiter docs:
```
// Wait is shorthand for WaitN(ctx, 1).
func (lim *Limiter) Wait(ctx context.Context) (err error) {
	return lim.WaitN(ctx, 1)
}

// WaitN blocks until lim permits n events to happen.
// It returns an error if n exceeds the Limiter's burst size, the Context is
// canceled, or the expected wait time exceeds the Context's Deadline.
// The burst limit is ignored if the rate limit is Inf.
func (lim *Limiter) WaitN(ctx context.Context, n int) (err error) { ...
```

Reply:
Ah, I misunderstood the Limiter, thought it was rejecting not forcing a wait. 👍